src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java

```diff
@@ -197,31 +196,70 @@ public synchronized CompactionTasks getUserDefinedTasks(Collection<? extends Com
+                // If possible, we want to issue separate compactions for non-overlapping sets of sstables, to allow
+                // for smaller extra space requirements. However, if the sharding configuration has changed, a major
+                // compaction should combine non-overlapping sets if they are split on a boundary that is no longer
+                // in effect.
+
+                // Select desired splitting of the sstables. If we are resharding, according to the sharding boundaries
+                // suitable for the size of the data. Otherwise, start with the overlap groups.
+                List<Set<CompactionSSTable>> groups =
+                    getShardManager().splitSSTablesInShards(arena.sstables,
+                                                            makeShardingStats(arena.sstables).shardCountForDensity,
+                                                            (sstableShard, shardRange) -> Sets.newHashSet(sstableShard));
+
+                // Now combine all of these groups that share an sstable so that we have valid independent transactions.
+                groups = combineSetsWithCommonElement(groups);
+
+                for (Collection<CompactionSSTable> set : groups)
+                {
+                    LifecycleTransaction txn = realm.tryModify(set, OperationType.COMPACTION);
+                    // Further split each of these into up to permittedParallelism tasks to run in parallel.
+                    if (txn != null)
+                        createAndAddTasks(gcBefore, txn, makeShardingStats(txn), permittedParallelism, tasks);
+                    // we ignore splitOutput (always split according to the strategy's sharding) and do not need isMaximal
+                }
             }
+
+            // If we have more arenas/non-overlapping sets than the permitted parallelism, we will try to run all the
+            // individual tasks in parallel (including as parallelized compactions) so that they finish quickest and release
+            // any space they hold, and then reuse the compaction thread to run the next set of tasks.
+            return CompactionTasks.create(CompositeCompactionTask.applyParallelismLimit(tasks, permittedParallelism));
         }
-        return CompactionTasks.create(tasks);
+        catch (Throwable t)
+        {
+            throw rejectTasks(tasks, t);
+        }
+    }
+
+    /**
+     * Split the given set of sstables into non-overlapping sets that split on boundaries that apply for the current
+     * sharding configuration.
+     */
+    public List<Set<CompactionSSTable>> splitInNonOverlappingShardSets(List<CompactionSSTable> sstables)
+    {
+        var outputShards = getShardManager().splitSSTablesInShards(sstables, makeShardingStats(sstables).shardCountForDensity,
+                                                                   (sstableShard, shardRange) -> new HashSet<>(sstableShard));
```

Comment:
This method is not used, and if we need it, it should be synchronized since `getShardManager()` reads/updates state.

Reply:
Looks like there are several calls to `getShardManager()` that probably need synchronization. Some of the usages were added in other PRs but appear incorrect, given the current pattern and usage of the method in this class.

Reply:
I missed the synchronization within the `maybeUpdateSelector()` method, please disregard the comments about synchronization on this thread.

We should still address the unused method.